EIPS/eip-2929.md

```diff
@@ -0,0 +1,165 @@
+| - | - | - |
+| `SLOAD_GAS` | 800 | `= WARM_STORAGE_READ_COST` |
+| `SSTORE_RESET_GAS` | 5000 | `5000 - COLD_SLOAD_COST` |
+
+The other parameters defined in EIP 2200 are unchanged.
+
+### SELFDESTRUCT changes
+
+If the ETH recipient of a `SELFDESTRUCT` is not in `accessed_addresses` (regardless of whether or not the amount sent is nonzero), charge an additional `COLD_ACCOUNT_ACCESS_COST` on top of the existing gas costs, and add the ETH recipient to the set.
+
+### Contract breakage mitigations
+
+See the Security Considerations section for details on the tradeoffs between these versions.
+
+#### Version 1
+
+Do nothing (ie. implement only the changes described above).
+
+#### Version 2
+
+Add a `POKE` precompile, with cost 4500 gas, that takes two stack arguments, `address` and `storage_slot`, and adds `address` to `accessed_addresses` and `(address, storage_slot)` to `accessed_storage_keys`.
+
+#### Version 3
+
+Access lists: see [EIP 2930](https://github.com/ethereum/EIPs/pull/2930)
+
+## Rationale
+
+### Opcode costs vs charging per byte of witness data
+
+The natural alternative path to changing gas costs to reflect witness sizes is to charge per byte of witness data. However, that would take a longer time to implement, hampering the goal of providing short-term security relief. Furthermore, following that path faithfully would lead to extremely high gas costs to transactions that touch contract code, as one would need to charge for all 24000 contract code bytes; this would be an unacceptably high burden on developers. It is better to wait for [code merklization](https://medium.com/ewasm/evm-bytecode-merklization-2a8366ab0c90) to start trying to properly account for gas costs of accessing individual chunks of code; from a short-term DoS prevention standpoint, accessing 24 kB from disk is not much more expensive than accessing 32 bytes from disk, so worrying about code size is not necessary.
+
+### Adding the accessed_addresses / accessed_storage_keys sets
+
+The sets of already-accessed accounts and storage slots are added to avoid needlessly charging for things that can be cached (and in all performant implementations already are cached). Additionally, it removes the current undesirable status quo where it is needlessly unaffordable to do self-calls or call precompiles, and enables contract breakage mitigations that involve pre-fetching some storage key allowing a future execution to still take the expected amount of gas.
+
+### SSTORE gas cost change
+
+The change to SSTORE is needed to avoid the possibility of a DoS attack that "pokes" a randomly chosen zero storage slot, changing it from 0 to 0 at a cost of 800 gas but requiring a de-facto storage load. The `SSTORE_RESET_GAS` reduction ensures that the total cost of SSTORE (which now requires paying the `COLD_SLOAD_COST`) remains unchanged. Additionally, note that applications that do `SLOAD` followed by `SSTORE` (eg. `storage_variable += x`) _would actually get cheaper_!
```

Comment:
```suggestion
The change to `SSTORE` is needed to avoid the possibility of a DoS attack that "pokes" a randomly chosen zero storage slot, changing it from 0 to 0 at a cost of 800 gas but requiring a de-facto storage load. The `SSTORE_RESET_GAS` reduction ensures that the total cost of `SSTORE` (which now requires paying the `COLD_SLOAD_COST`) remains unchanged. Additionally, note that applications that do `SLOAD` followed by `SSTORE` (eg. `storage_variable += x`) _would actually get cheaper_!
```